Reimplement Workbench integration as addon

matt-diehl opened the pull request:
@MarkusBordihn, @adamcarheden, @MartenvanWezel 

This PR is a reimplementation of the CwC/Workbench integration from #132 that keeps all of the logic in an addon. Two demo projects can be found in the Beginner > Blockly section, after enabling experimental mode.

Some notes:

1. This is currently pulling projects from a staging site for testing purposes.
2. Workbench projects are appended to a given section if the project has a matching tag. I'm appending the projects to the current "tutorials" section, since I wasn't sure if there was sense in distinguishing Workbench projects from the other tutorials on the site, and, if we did, how to distinguish them. If we go with this approach, every section in CwC will need that section for this addon to append to (it currently searches for a div with class "cwc-tutorial-list" in each section).
3. To utilize the same tutorial button on the sidebar would have required changes to the sidebar component, so I instead added another section via the sidebar's ```addCustomButton``` method. I'm not sure how you all feel about this. Perhaps this is an alright approach, but in the projects listing, we should put Workbench projects in their own section?
4. I didn't dig into it too much, but the custom sidebar button tooltip isn't working for me.
5. I used the new database utility for storing projects and project images - is this the correct usage? I saw there's also a cache utility, but I wasn't clear on when that one should be used. Also, I added a delete method to the database utility.
6. The app is missing a "mode" for the Sphero BB8 robot.
7. I created a media overlay that I think is better suited to images, and is in line with the Material Design guidelines, instead of attempting to utilize the dialog component. It is all self-contained in the addon. Let me know if you disagree with this approach and I can rework it.
8. I updated the ```loadMode``` method in src/mode/modder.js to return a promise so I could use the sidebar after it was rendered.

MartenvanWezel commented:
Awesome - thanks Matt. Of course I'll let the eng team review. Just a few questions:

1/ Have you implemented a graceful catch on problems, in particular:
   a/ Connection trouble when trying to fetch the very first projects
   b/ Connection trouble when trying to refresh
   c/ Broken content from Workbench. (corrupted project data somehow..)
2/ How many kb/mb is one project typically? 
3/ Is it theoretically straightforward to sideload projects?
4/ Did you implement the switch in settings to turn off the fetching of WB projects?
5/ Are you drawing something somewhere that indicates that background download is going on, with a nice tooltip ("downloading 'Build a voice controlled Sphero' project..")

matt-diehl commented:
1a. Currently if the initial request fails, no projects are added the database, and therefore no Workbench projects are visible to the user. Do you want me to present a notification to the user if it fails?
1b. Subsequent loads check for new or modified projects and attempts to load them. The same logic applies: if the project fails to download, do not add it (or delete it in the case of a modified project). I haven't implemented a refresh functionality; projects only download when the app is opened. Is this something you'd like to see added? I'd need to confer with the others about where the best place to add that would be, without modifying core files.
1c. I'll think about the best way to verify the project data is not corrupt and implement something.
2. The project text content as downloaded from Workbench will vary based on the number of steps, but 3-6kb is probably average. Images will of course be a lot more but that depends entirely on how image heavy the project is. YouTube videos aren't downloaded and only shown as available at runtime if the user has an internet connection.
3. I'm not totally clear on what you're asking, but if it's in regards to being simple to publish, yes. We will give permissions to you or anyone else who needs it to publish directly to a "Coding with Chrome" channel on the Workbench site. Depending on the tags you assign, the project will show up in the relevant section the next time the app is loaded.
4. No, I'll need @MarkusBordihn to comment on how he'd like to see this implemented without modifying the core files - should we add methods for adding settings options?
5. No, but I could use the existing bottom alert to do this. What is the main goal of it? Just to indicate that background downloading is going on, or to let the user know there's new content? That UI element only shows for a second before going away, so I'm not sure how much information the user could get from it.

MarkusBordihn commented on src/addon/workbench/workbench.js:
```diff
@@ -0,0 +1,238 @@
+  /** @type {!cwc.utils.Helper} */
+  this.helper = helper;
+
+  /** @private {!cwc.utils.Database} */
```
Would it be possible to use an single Database('Workbench') with two different ObjectStoreName like:
+  this.projectsDb_ = new cwc.utils.Database('Workbench')
+    .setObjectStoreName('__workbench_projects__');
+
+  /** @private {!cwc.utils.Database} */
+  this.imagesDb_ = new cwc.utils.Database('Workbench')
+    .setObjectStoreName('__workbench_images__');

matt-diehl replied:
This is now updated. I'm not sure if it's how it's expected to work, but I had to pass the array of objectStoreNames when opening each database, or else the images database would not work properly.

MarkusBordihn commented on src/addon/workbench/workbench_loader.js:
```diff
@@ -0,0 +1,186 @@
+  if (!this.helper.checkFeature('online')) return;
+
+  let xhr = new goog.net.XhrIo();
+  goog.events.listen(xhr, goog.net.EventType.SUCCESS, (e) => {
```
Would be great if you could add getUriAsJson to cwc.utils.Resources and use the function from here.

MarkusBordihn commented:
Additional comments:
* For the settings screen we need to provide some API function and an additional tab for the Addon's.
* Indicator for Background loading is not really needed.
As soon an mode is loaded there is already an indicator to showing the loading status but most of the time the user is not seeing it, so it will only show up in case there is an error and the mode could not be loaded.

MartenvanWezel commented:
1a> No that's fine, no notification needed
1b> Hmm, I'm not too sure about how often CWC is restarted if it is a primary use for a machine. Perhaps it makes sense to occasionally refresh but let's punt that to "v1.1" unless it's easy
1c> thanks!

2+3> Context: CwC is very low network footprint and I would like to support a very internet-poor use case, e.g. a teacher would "download lesson plans" from his local internet cafe, and then distribute them through USB sticks or something along those lines. 
Hence the question if we can download, sideload, offline distribute.. 

4> I'll let you  handle settings screen, thanks.
5> Basically a similar point to 4, but I prefer people to be aware that "stuff they might want to know about" is happening. Given how small the data is (10kb?) it might indeed flit across the screen too quickly to be helpful... 

How about this:
https://github.com/google/coding-with-chrome/issues/165

adamcarheden commented:
@matt-diehl  - This looks great!

@MarkusBordihn and @MartenvanWezel - There are a few differences between Blockly between CwC and Workbench. (i.e. Our "Command" tab is called "General" on Workbench). It probably doesn't need to be done as part of this PR, but would it be helpful to sync them up? Otherwise project creators may write instructions that match up with the UI they see on Workbench but confuse CwC users because they don't match up with the CwC Blockly UI.

I'm glag to submit a PR to make the CwC blockly tabs match the ones on Workbench.

adamcarheden commented:
Looks like I spoke too soon. @matt-diehl pointed out that CwC supports the Blockly drawing tools, but Workbench does not, so it's likely more complicated than changing the order of things and some text labels.

I'm still glad to sync the Blockly tabs to the extent possible if anyone thinks that would be helpful though.

MartenvanWezel commented:
I'll let @MarkusBordihn comment authoritatively but as I understand it this is a 'hard' problem to solve, since the toolbox of blockly is fairly open-to-interpretation, i.e. in the official docs - [Toolbox](https://developers.google.com/blockly/guides/configure/web/toolbox) - there is neither a 'general' nor a 'command' category defined. 

Frankly, as far as I'm aware, our blockly toolboxes overall have been generated in a fairly 'ad-hoc' way rather than be thought in deep detail. 

I suspect that if we change them today, old blockly code will stop working. This is not fun, _but_ might be good to do sooner rather than later. 

@matt-diehl - does your company have some official, well-thought-out strategy around your use of blockly, consistency of toolboxes across your various products/customers etc?

@MarkusBordihn - is there some way we can provide backwards compatibility with our old blockly toolbox layout? Or will it work anyway?

matt-diehl commented:
@MartenvanWezel, I don't think it's a realistic goal to make our two programming canvases aligned 100%. They share most of the basics, but the Workbench Blockly instance is designed around supporting multi-device programming, which is not part of CwC, and there are features that CwC has that Workbench may or may not support, such as a drawing canvas. I think it just needs to be clear that a project is designed to be performed in CwC.

MartenvanWezel commented:
Thanks @matt-diehl - makes sense.

Still, while agreed that trying to be verbatim copy is unrealistic, CwC could use a bit of a blockly-overhaul to make it more consistent across the product. Will kick off a minor effort elsewhere, but if you're willing to share your company's guidelines, if any exist, that would be helpful.

matt-diehl commented:
@MartenvanWezel, we don't have any written guidelines on how we organize our Blockly blocks. We're testing in schools and getting positive feedback, but it's subject to change as we test with more users.

I'll make the changes requested by Markus in the next day or two.

@MarkusBordihn - will you be adding the API for adding settings, or is this something you'd like me to work on?

matt-diehl commented:
I made the two changes requested by Markus, added a check for data to make sure it does not appear corrupt, and added a setting (hardcoded) under "misc" to toggle the auto-fetching of new content from Workbench. I also updated this to point to the production Workbench API.

The only remaining bit, I believe, is to add the tutorial dom nodes to each section that this addon can use to append projects to. However, the more I looked at that, the more the technical implementation seemed like it would be a bit messy -- each tutorials section has a heading, but the heading should only display if there are either tutorials from either CwC or Workbench. So, something needs to control that display, but I'm not sure that job makes sense in this addon, nor in the select screen.

So the options are:

1. Stick with this approach where Workbench and CwC tutorials are mixed in the same area. The tutorials container/title will always exist in the template, but be hidden by default. If there are CwC tutorials, CwC will show the container. If there are Workbench projects, the addon will show the container.
2. Workbench projects get their own "Workbench Projects" section, which the addon appends, and removes the need for any template additions. This is the direction I'm leaning now.

@MarkusBordihn / @adamcarheden - let me know your thoughts on that.

adamcarheden commented:
I'm fine with #2.

On Tue, May 29, 2018, 2:42 PM Matt <notifications@github.com> wrote:

> I made the two changes requested by Markus, added a check for data to make
> sure it does not appear corrupt, and added a setting (hardcoded) under
> "misc" to toggle the auto-fetching of new content from Workbench. I also
> updated this to point to the production Workbench API.
>
> The only remaining bit, I believe, is to add the tutorial dom nodes to
> each section that this addon can use to append projects to. However, the
> more I looked at that, the more the technical implementation seemed like it
> would be a bit messy -- each tutorials section has a heading, but the
> heading should only display if there are either tutorials from either CwC
> or Workbench. So, something needs to control that display, but I'm not sure
> that job makes sense in this addon, nor in the select screen.
>
> So the options are:
>
>    1. Stick with this approach where Workbench and CwC tutorials are
>    mixed in the same area. The tutorials container/title will always exist in
>    the template, but be hidden by default. If there are CwC tutorials, CwC
>    will show the container. If there are Workbench projects, the addon will
>    show the container.
>    2. Workbench projects get their own "Workbench Projects" section,
>    which the addon appends, and removes the need for any template additions.
>    This is the direction I'm leaning now.
>
> @MarkusBordihn <https://github.com/MarkusBordihn> / @adamcarheden
> <https://github.com/adamcarheden> - let me know your thoughts on that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/coding-with-chrome/pull/164#issuecomment-392936783>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABAqBOBvlMIP-OpUyLDFDnuKa-tDuaceks5t3bKygaJpZM4UJFMW>
> .
>

matt-diehl commented:
@MarkusBordihn, should I create a new PR for the change I mentioned in my above comment?

MarkusBordihn commented:
@matt-diehl
Yes, please. Feel free to add any additional images, logos for this separate section, as long you have the needed image rights and they fall under our currently used license.
